src/components/device/smartbase.ts

```
@@ -0,0 +1,226 @@
+          } else {
+            Logger.error('Failed performing a camera upgrade', e, 'Smartbase API');
+            reject(e);
+          }
+        }
+      };
+      tryRunAgainOnFailure(this);
+    });
+  }
+
+  async createAndRunUpgrade(
+    opts: UpgradeOpts,
+    deviceManager: IDeviceManager,
+    createNewUpgrader: boolean
+  ) {
+    let upgrader: IDeviceUpgrader = opts.upgrader;
+    if (!upgrader || createNewUpgrader) {
+      upgrader = await createBoxfishUpgrader(deviceManager, this.discoveryEmitter);
+    }
+    upgrader.init(opts);
+    upgrader.start();
+    return new Promise<void>((resolve, reject) => {
+      upgrader.once(CameraEvents.UPGRADE_COMPLETE, async (deviceManager) => {
+        const upgradeIsOk = await upgrader.upgradeIsValid();
+        if (upgradeIsOk) {
+          resolve();
+        } else {
+          reject({
+            message: 'Upgrade status is not ok, run again',
+            runAgain: true,
+            deviceManager,
+          });
+        }
+      });
+      upgrader.once(CameraEvents.UPGRADE_FAILED, (reason) => {
+        Logger.error('Upgrade Failed', reason, 'Boxfish API');
+        reject(reason);
+      });
+      upgrader.once(CameraEvents.TIMEOUT, (reason) => {
+        Logger.error('Upgrader returned a timeout event', reason, 'Boxfish API');
+        reject(reason);
+      });
+    });
+  }
+
+  getAutozoomControl(opts: AutozoomControlOpts): ICnnControl {
+    throw new Error('Not supported for Smartbase.');
+  }
+
+  getFaceBasedExposureControl(): ICnnControl {
+    throw new Error('Not supported for Smartbase.');
+  }
+
+  getDetector(opts?: DetectorOpts): IDetector {
+    throw new Error('Not supported for Smartbase.');
+  }
+
+  getDiagnostics(): Promise<DiagnosticsMessage[]> {
+    throw new Error('Not supported for Smartbase.');
+  }
+  async getState(): Promise<any> {
+    const { autozoom_enabled } = await this.api.getProductInfo();
+    return { autozoom_enabled };
```

Throw error instead of returning this imo. It doesn't mean anything